### PR TITLE
Start moving out some parts into the new `grammer` crate.

### PR DESCRIPTION
**Pull request by eddyb**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -10,7 +10,8 @@ mod indexing_str;
 #[forbid(unsafe_code)]
 pub mod generate;
 #[forbid(unsafe_code)]
-pub mod grammar;
+pub mod parse_node;
+#[forbid(unsafe_code)]
```
Comment (qmx) [inline src/lib.rs]: leftover?

Comment (eddyb) [reply src/lib.rs]: Wow, `kdiff3` confused me into this.

File: build.rs [hunk 2]
```diff
@@ -6,123 +7,27 @@ extern crate proc_quote;
...
-                {Many:"*"} |
-                {More:"+"};
-            Pattern =
-                {Str:StrLit} |
-                {CharRange:{ {{start:CharLit}?} ".." {{end:CharLit}?} }} |
-                {CharRangeInclusive:{ {{start:CharLit}?} "..=" {end:CharLit} }};
-        }
-        .insert_whitespace(grammar::call("Whitespace")),
-    );
+    let grammar: Grammar = grammer::grammar_grammar();
```
Comment (Centril) [inline build.rs]: This is pretty ungreat...

File: src/generate/rust.rs [hunk 3]
```diff
@@ -459,12 +485,27 @@ impl Continuation<'_> {
...
 impl<F: FnOnce(Continuation<'_>) -> Continuation<'_>> ContFn for F {
     fn apply(self, cont: Continuation<'_>) -> Continuation<'_> {
         self(cont)
     }
+    fn apply_box(self: Box<Self>, cont: Continuation<'_>) -> Continuation<'_> {
+        (*self).apply(cont)
+    }
+}
+
+impl<'a> ContFn for Box<dyn ContFn + 'a> {
```
Comment (Centril) [inline src/generate/rust.rs]: You should be able to elide the lifetime in the header here?

Comment (eddyb) [reply src/generate/rust.rs]: Argh, I forgot about that! I think I did use `'_` the first attempt.

**Pull request merged**